fem/transfer.cpp

```diff
@@ -543,17 +543,172 @@ void L2ProjectionGridTransfer::L2ProjectionL2Space::ProlongateTranspose(
 }
 
 L2ProjectionGridTransfer::L2ProjectionH1Space::L2ProjectionH1Space(
-   const FiniteElementSpace& fes_ho_, const FiniteElementSpace& fes_lor_)
-   : L2Projection(fes_ho_, fes_lor_)
+   const FiniteElementSpace& fes_ho_, const FiniteElementSpace& fes_lor_,
+   bool build_operators)
+   : L2Projection(fes_ho_, fes_lor_),
+#ifdef MFEM_USE_MPI
+     pcg { dynamic_cast<const ParFiniteElementSpace*>(&fes_ho) ?
+           CGSolver(MPI_COMM_WORLD) : CGSolver()
+         }
+#else
+     pcg()
+#endif
 {
+   if (build_operators)
+   {
+      std::tie(R, M_LH) = ComputeSparseRAndM_LH();
+      RTxM_LH = std::unique_ptr<Operator>(TransposeMult(static_cast<SparseMatrix&>
+                                                        (*R),
+                                                        static_cast<SparseMatrix&>(*M_LH)));
+   }
+
+   // Basic PCG solver setup
+   pcg.SetPrintLevel(0);
+   pcg.SetMaxIter(1000);
+   // initial values for relative and absolute tolerance
+   SetRelTol(1e-13);
+   SetAbsTol(1e-13);
+   if (build_operators)
+   {
+      precon = std::unique_ptr<DSmoother>(new DSmoother(static_cast<SparseMatrix&>
+                                                        (*RTxM_LH)));
+      pcg.SetPreconditioner(*precon);
+      pcg.SetOperator(*RTxM_LH);
+   }
+}
+
+void L2ProjectionGridTransfer::L2ProjectionH1Space::Mult(
+   const Vector& x, Vector& y) const
+{
+   Vector X(fes_ho.GetTrueVSize());
+   Vector X_dim(R->Width());
+
+   Vector Y_dim(R->Height());
+   Vector Y(fes_lor.GetTrueVSize());
+
+   Array<int> vdofs_list;
+
+   GetTDofs(fes_ho, x, X);
+
+   for (int d = 0; d < fes_ho.GetVDim(); ++d)
+   {
+      TDofsListByVDim(fes_ho, d, vdofs_list);
+      X.GetSubVector(vdofs_list, X_dim);
+      R->Mult(X_dim, Y_dim);
+      TDofsListByVDim(fes_lor, d, vdofs_list);
+      Y.SetSubVector(vdofs_list, Y_dim);
+   }
+
+   SetTDofs(fes_lor, std::move(Y), y);
+}
+
+void L2ProjectionGridTransfer::L2ProjectionH1Space::MultTranspose(
+   const Vector& x, Vector& y) const
+{
+   Vector X(fes_lor.GetTrueVSize());
+   Vector X_dim(R->Height());
+
+   Vector Y_dim(R->Width());
+   Vector Y(fes_ho.GetTrueVSize());
+
+   Array<int> vdofs_list;
+
+   GetTDofs(fes_lor, x, X);
+
+   for (int d = 0; d < fes_ho.GetVDim(); ++d)
+   {
+      TDofsListByVDim(fes_lor, d, vdofs_list);
+      X.GetSubVector(vdofs_list, X_dim);
+      R->MultTranspose(X_dim, Y_dim);
+      TDofsListByVDim(fes_ho, d, vdofs_list);
+      Y.SetSubVector(vdofs_list, Y_dim);
+   }
+
+   SetTDofs(fes_ho, std::move(Y), y);
+}
```

Comment:
I don't think the action of `MultTranspose` is correct when the prolongation and restriction matrices are non-trivial.

To be specific, `GetTDofs` corresponds to multiplication by `R`, and `SetTDofs` corresponds to multiplication by `P`. So in the transpose, the sequence should be `P^T` at the beginning and `R^T` at the end.

(Also, `SetTDofs` should probably be called `SetFromTDofs`).

Reply:
Thanks for noticing this!  It's been fixed in the latest commit.